junit-platform-engine/src/main/java/org/junit/platform/engine/TestDescriptor.java

```diff
@@ -172,6 +176,30 @@ default Set<? extends TestDescriptor> getDescendants() {
 	 */
 	void removeFromHierarchy();
 
+	/**
+	 * Order all children from this descriptor.
+	 *
+	 * <p>The {@code orderer} is provided a modifiable list of children in
+	 * this test descriptor. Never {@code null}. The {@code orderer} may return
+	 * the elements of list in any order but may not add or remove descriptors.
+	 * <p>
+	 * If descriptors were added or removed this method must throw a
+	 * {@link org.junit.platform.commons.JUnitException JUnitException} explaining
+	 * that an {@code orderer} may not remove children from this descriptor.
+	 *
+	 * @param orderer a unary operator to order the children from this test descriptor.
+	 */
+	@API(since = "5.12", status = EXPERIMENTAL)
+	default void orderChildren(UnaryOperator<List<TestDescriptor>> orderer) {
+		Preconditions.notNull(orderer, "orderer must not be null");
+		Set<? extends TestDescriptor> children = getChildren();
+		List<? extends TestDescriptor> orderedChildren = orderer.apply(new ArrayList<>(children));
+		boolean unmodified = children.size() == orderedChildren.size() && children.containsAll(orderedChildren);
+		Preconditions.condition(unmodified, "orderer may not add or remove test descriptors");
+		orderedChildren.forEach(this::removeChild);
+		orderedChildren.forEach(this::addChild);
```

Comment:
This could be optimized by pushing down the implementation, but then the interface still needs a sensible default.

Reply:
I think having this unoptimized version in the interface is fine but we should have an optimized version in `AbstractTestDescriptor`